// @vtex/api/lib/clients/IOClients.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = require("./index");
class IOClients {
    constructor(clientOptions, ctx) {
        this.clientOptions = clientOptions;
        this.ctx = ctx;
        this.clients = {};
    }
    get apps() {
        return this.getOrSet('apps', index_1.Apps);
    }
    get billing() {
        return this.getOrSet('billing', index_1.Billing);
    }
    get billingMetrics() {
        return this.getOrSet('billingMetrics', index_1.BillingMetrics);
    }
    get builder() {
        return this.getOrSet('builder', index_1.Builder);
    }
    get events() {
        return this.getOrSet('events', index_1.Events);
    }
    get id() {
        return this.getOrSet('id', index_1.ID);
    }
    get licenseManager() {
        return this.getOrSet('licenseManager', index_1.LicenseManager);
    }
    get logger() {
        return this.getOrSet('logger', index_1.Logger);
    }
    get messages() {
        return this.getOrSet('messages', index_1.Messages);
    }
    get metadata() {
        return this.getOrSet('metadata', index_1.Metadata);
    }
    get registry() {
        return this.getOrSet('registry', index_1.Registry);
    }
    get router() {
        return this.getOrSet('router', index_1.Router);
    }
    get segment() {
        return this.getOrSet('segment', index_1.Segment);
    }
    get settings() {
        return this.getOrSet('settings', index_1.Settings);
    }
    get session() {
        return this.getOrSet('session', index_1.Session);
    }
    get vbase() {
        return this.getOrSet('vbase', index_1.VBase);
    }
    get workspaces() {
        return this.getOrSet('workspaces', index_1.Workspaces);
    }
    getOrSet(key, Implementation) {
        const options = Object.assign({}, this.clientOptions.default, this.clientOptions[key], { metrics });
        if (!this.clients[key]) {
            this.clients[key] = new Implementation(this.ctx, options);
        }
        return this.clients[key];
    }
}
exports.IOClients = IOClients;
