// @vtex/api/lib/clients/LicenseManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const qs_1 = require("qs");
const HttpClient_1 = require("../HttpClient");
const TWO_MINUTES_S = 2 * 60;
const routes = {
    accountData: '/api/license-manager/account',
    topbarData: '/api/license-manager/site/pvt/newtopbar',
};
const inflightKey = ({ baseURL, url, params }) => {
    return baseURL + url + qs_1.stringify(params, { arrayFormat: 'repeat', addQueryPrefix: true });
};
class LicenseManager extends HttpClient_1.JanusClient {
    getAccountData() {
        const { authToken } = this.context;
        return this.http.get(routes.accountData, {
            forceMaxAge: TWO_MINUTES_S,
            headers: {
                VtexIdclientAutCookie: authToken,
            },
            inflightKey,
            metric: 'lm-account-data',
        });
    }
    getTopbarData(VtexIdclientAutCookie) {
        return this.http.get(routes.topbarData, {
            headers: {
                VtexIdclientAutCookie,
            },
            metric: 'lm-topbar-data',
        });
    }
}
exports.LicenseManager = LicenseManager;
