// @vtex/api/lib/clients/Segment.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const cookie_1 = __importDefault(require("cookie"));
const ramda_1 = require("ramda");
const HttpClient_1 = require("../HttpClient");
const SEGMENT_COOKIE = 'vtex_segment';
const SEGMENT_MAX_AGE_S = 10 * 60; // 10 minutes - segment is actually immutable
const sanitizeParams = (params) => {
    return ramda_1.pickBy((_, key) => !!key, params || {});
};
const routes = {
    base: '/api/segments',
    segments: (token) => token ? `${routes.base}/${token}` : routes.base,
};
class Segment extends HttpClient_1.JanusClient {
    constructor() {
        super(...arguments);
        /**
         * @deprecated Please use `getSegment` or `getSegmentByToken` instead.
         *
         * @memberof Segment
         */
        this.segment = (query, token) => this.rawSegment(token || this.context.segmentToken, query).then(ramda_1.prop('data'));
        /**
         * Get the segment data using the current `ctx.vtex.segmentToken`
         *
         * @memberof Segment
         */
        this.getSegment = () => this.rawSegment(this.context.segmentToken).then(ramda_1.prop('data'));
        /**
         * Get the segment data from this specific segment token
         *
         * @memberof Segment
         */
        this.getSegmentByToken = (token) => this.rawSegment(token).then(ramda_1.prop('data'));
        this.getOrCreateSegment = async (query, token) => {
            const { data: segmentData, headers: { 'set-cookie': [setCookies], }, } = await this.rawSegment(token, query);
            const parsedCookie = cookie_1.default.parse(setCookies);
            const segmentToken = ramda_1.prop(SEGMENT_COOKIE, parsedCookie);
            return {
                segmentData,
                segmentToken,
            };
        };
        this.rawSegment = (token, query) => {
            return this.http.getRaw(routes.segments(token), ({
                forceMaxAge: SEGMENT_MAX_AGE_S,
                headers: {
                    'Content-Type': 'application/json',
                },
                inflightKey: HttpClient_1.inflightUrlWithQuery,
                metric: 'segment-get',
                params: Object.assign({}, sanitizeParams(query)),
            }));
        };
    }
}
exports.Segment = Segment;
